telemetry/event/event.go

```diff
@@ -0,0 +1,97 @@
+package event
+
+import (
+	"errors"
+	"runtime"
+	"time"
+
+	"go.dagger.io/dagger/version"
+)
+
+var (
+	ErrMalformedEvent = errors.New("malformed event properties")
+)
+
+const eventVersion = "2022-05-25.01"
+
+type Event struct {
+	Name string `json:"name"`
+
+	Version   string    `json:"v"`
+	Timestamp time.Time `json:"ts"`
+
+	Data Properties `json:"data"`
+
+	Engine engineProperties `json:"engine"`
+}
+
+type engineProperties struct {
+	ID string `json:"id"`
+
+	Version  string `json:"version"`
+	Revision string `json:"revision"`
+
+	OS   string `json:"os"`
+	Arch string `json:"arch"`
+}
+
+func (e engineProperties) Validate() error {
+	switch {
+	// FIXME: not implemented
+	// case e.ID == "":
+	// 	return ErrMalformedEvent
+	case e.Version == "":
+		return ErrMalformedEvent
+	case e.OS == "":
+		return ErrMalformedEvent
+	case e.Arch == "":
+		return ErrMalformedEvent
+	}
+	return nil
+}
+
+type Properties interface {
+	EventName() string
+	EventVersion() string
+	Validate() error
+}
+
+func (e *Event) Validate() error {
+	switch {
+	case e.Name == "":
+		return ErrMalformedEvent
+	case e.Version == "":
```

Comment:
I would also validate the string format of the `version`. I noticed some with `v0.2.12` and `0.2.12`.